src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs

```diff
@@ -279,12 +285,118 @@ protected virtual bool IsNewHeadAlignedWithChain(Block newHeadBlock, ForkchoiceS
+        const int maxOutputChars = 32 * 2;
+
+        if (bytes == null || bytes.Length == 0 || bytes.Length > 32)
+            return string.Empty;
+
+        // Allocate a char buffer on the stack.
+        Span<char> outputBuffer = stackalloc char[maxOutputChars];
+
+        int outputPos = 0;
+        int index = 0;
+        bool hasValidContent = false;
+        bool shouldAddSpace = false;
+
+        while (index < bytes.Length)
+        {
+            ReadOnlySpan<byte> span = bytes.AsSpan(index);
+
+            OperationStatus status = Rune.DecodeFromUtf8(span, out Rune rune, out var bytesConsumed);
+            if (status == OperationStatus.Done)
+            {
+                if (!IsControlCharacter(rune))
+                {
+                    if (shouldAddSpace)
+                    {
+                        outputBuffer[outputPos++] = ' ';
+                        shouldAddSpace = false;
+                    }
+
+                    int charsNeeded = rune.Utf16SequenceLength;
+                    if (outputPos + charsNeeded > outputBuffer.Length)
+                    {
+                        // Expand output buffer
+                        int newSize = outputBuffer.Length * 2;
+                        char[] newBuffer = new char[newSize];
+                        outputBuffer.Slice(0, outputPos).CopyTo(newBuffer);
+                        outputBuffer = newBuffer;
+                    }
+
+                    rune.EncodeToUtf16(outputBuffer.Slice(outputPos));
+                    outputPos += charsNeeded;
+                    hasValidContent = true;
+                }
+                else
+                {
+                    // Control character encountered; set flag to add space if needed
+                    if (hasValidContent)
+                        shouldAddSpace = true;
+                }
+                index += bytesConsumed;
+            }
+            else if (status == OperationStatus.InvalidData)
+            {
+                // Invalid data; set flag to add space if needed
+                if (hasValidContent)
+                    shouldAddSpace = true;
+                index++;
+            }
+            else if (status == OperationStatus.NeedMoreData)
+            {
+                // Incomplete sequence at the end; break out of the loop
+                break;
+            }
+            else
+            {
+                // Unexpected status; treat as invalid data
+                if (hasValidContent)
+                    shouldAddSpace = true;
+                index++;
+            }
+        }
+
+        // Create the final string from the output buffer.
+        return outputPos > 0 ? new string(outputBuffer[..outputPos]) : string.Empty;
+    }
+
+    private static bool IsControlCharacter(Rune rune)
+    {
+        // Control characters are U+0000 to U+001F and U+007F to U+009F
+        return rune.Value <= 0x001F || (rune.Value >= 0x007F && rune.Value <= 0x009F);
+    }
```

Comment:
Again sounds a bit out of place in `ForkChoiceUpdateHandler`

Reply:
Moved to Byte extensions; also moved ExtraData from ForkChoice to NewPayload